recipes/nasm/all/conanfile.py

```diff
@@ -11,85 +18,100 @@ class NASMConan(ConanFile):
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
 
     def package_id(self):
         del self.info.settings.compiler
 
-    def _build_vs(self):
-        with tools.chdir(self._source_subfolder):
-            with tools.vcvars(self):
-                autotools = AutoToolsBuildEnvironment(self)
-                autotools.flags.append("-nologo")
-                self.run("nmake /f {} {}".format(os.path.join("Mkfiles", "msvc.mak"), " ".join("{}=\"{}\"".format(k, v) for k, v in autotools.vars.items())))
-                shutil.copy("nasm.exe", "nasmw.exe")
-                shutil.copy("ndisasm.exe", "ndisasmw.exe")
-
-    def _configure_autotools(self):
-        if self._autotools:
-            return self._autotools
-        self._autotools = AutoToolsBuildEnvironment(self, win_bash=(self._settings_build.os == "Windows"))
+    def generate(self):
+        tc = AutotoolsToolchain(self)
+        if is_msvc(self):
+            VCVars(self).generate()
+            tc.configure_args.append("-nologo")
         if self.settings.arch == "x86":
-            self._autotools.flags.append("-m32")
+            tc.extra_cflags.append("-m32")
         elif self.settings.arch == "x86_64":
-            self._autotools.flags.append("-m64")
-        self._autotools.configure(configure_dir=self._source_subfolder)
+            tc.extra_cflags.append("-m64")
+        tc.generate()
 
-        # GCC9 - "pure" attribute on function returning "void"
-        tools.replace_in_file("Makefile", "-Werror=attributes", "")
+        env = VirtualBuildEnv(self)
+        env.generate()
 
-        # Need "-arch" flag for the linker when cross-compiling.
-        # FIXME: Revisit after https://github.com/conan-io/conan/issues/9069, using new Autotools integration
-        if str(self.version).startswith("2.13"):
-            tools.replace_in_file("Makefile", "$(CC) $(LDFLAGS) -o", "$(CC) $(ALL_CFLAGS) $(LDFLAGS) -o")
+        # TODO I assume this is required
+        # inject requires env vars in build scope
+        # it's required in case of native build when there is AutotoolsDeps & at least one dependency which might be shared, because configure tries to run a test executable
+        if not cross_building(self):
+            env = VirtualRunEnv(self)
+            env.generate(scope="build")
 
-        return self._autotools
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        if self.settings.compiler == "Visual Studio":
-            self._build_vs()
+        apply_conandata_patches(self)
+
+        if is_msvc(self):
+            with chdir(self, self.source_folder):
+                self.run("nmake /f {}".format(os.path.join("Mkfiles", "msvc.mak")))
+                shutil.copy("nasm.exe", "nasmw.exe")
+                shutil.copy("ndisasm.exe", "ndisasmw.exe")
```

Comment:
please, use conan copy. It's not only related to copying only one file, but preserves metadata and deal with some python exceptions

Reply:
Sorry, I don't get it.
How do I copy a file to a new name?
conan.tools.files.copy() copies a file to another dir using the same name ...

Reply:
Ah yes, for your case `files.copy()` will not work because you need to rename. In that case, prefer `shutil.copy2()` instead. It preserves file's metadata, it's same idea used by `files.copy()`